### PR TITLE
In messages sockets

### PR DESCRIPTION
**Pull request by gursharan4312**
### What this PR does (required):
- Integrate sockets to messaging
- Add online/offline users
- Sends messages when user is online
- Typing animation when other user is typing

### Screenshots / Videos (required):

https://user-images.githubusercontent.com/35828532/121118055-0b32c880-c7ce-11eb-9533-a6ac4ce2ed33.mp4

### CODE REVIEW

**Comment by gursharan4312** [comment]
New Changes:

- Implemented read status
- Made UI responsive

https://user-images.githubusercontent.com/35828532/121253546-159ca300-c85e-11eb-9701-7b1cd28c9cd2.mp4

File: client/src/context/useMessageContext.tsx [hunk 1]
```diff
@@ -55,53 +94,103 @@ export const MessageContextProvider: FunctionComponent = ({ children }): JSX.Ele
...
     },
     [history, conversations],
   );
 
+  const addNewMessage = (message: Message) => {
+    setConversations((conversations) => {
+      //if message is already added
+      const convo = conversations.find((convo) => convo.conversationId === message.conversationId);
+      if (convo?.messages && convo.messages.find((msg) => msg._id === message._id)) {
+        console.log('already present');
```
Comment (xia0m) [inline client/src/context/useMessageContext.tsx]: Remove

File: client/src/context/useMessageContext.tsx [hunk 2]
```diff
@@ -55,53 +94,103 @@ export const MessageContextProvider: FunctionComponent = ({ children }): JSX.Ele
...
+        return conversations;
+      }
+      // add messages to conversation
+      console.log('adding present');
```
Comment (xia0m) [inline client/src/context/useMessageContext.tsx]: remove

File: client/src/pages/Messages/ActiveChat/ActiveChat.tsx [hunk 3]
```diff
@@ -13,22 +13,36 @@ interface Props {
   handleSendMessage: (text: string) => void;
 }
 
-const SideBar = ({ conversation, handleSendMessage }: Props): JSX.Element => {
+const ActiceChat = ({ conversation, handleSendMessage }: Props): JSX.Element => {
   const classes = useStyles();
   const { loggedInUserDetails } = useAuth();
   const { loading, messages } = useMessages();
+  // if (conversation?.recipient.typing) console.log('typing from message');
```
Comment (xia0m) [inline client/src/pages/Messages/ActiveChat/ActiveChat.tsx]: remove

File: client/src/index.css [hunk 4]
```diff
@@ -1,13 +1,53 @@
...
+  0% {
+    top: 0rem;
+  }
+  25% {
+    top: 0.5rem;
+  }
+  100% {
+    top: -0.5rem;
+  }
 }
```
Comment (xia0m) [inline client/src/index.css]: is there any reason the style for a component is defined in global CSS file? can you put the style in `otherUserBubble.tsx`? The style in `useStyles()` is provided below
```
typingLoader: {
    position: 'relative',
    height: '2rem',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    margin: 'auto',
    padding: '0 0.5rem',
    '& div': {
      display: 'inline-block',
      width: '0.6rem',
      height: '0.6rem',
      borderRadius: '50%',
      background: '#ddd',
      margin: '0 0.3rem',
      animation: 'typing-loader-animation 1.2s cubic-bezier(0, 0.5, 0.5, 1) infinite',
      position: 'relative',
      '&:nth-child(1)': {
        animationDelay: '-0.24s',
      },
      '&:nth-child(2)': {
        animationDelay: '-0.12s',
      },
      '&:nth-child(3)': {
        animationDelay: '-0',
      },
    },
  },
  '@keyframes typing-loader-animation': {
    '0%': {
      top: '0rem',
    },
    '25%': {
      top: '0.5rem',
    },
    '100%': {
      top: '-0.5rem',
    },
  },
```
```
<div className={classes.typingLoader}>
  <div></div>
  <div></div>
  <div></div>
</div>
```

Comment (gursharan4312) [reply client/src/index.css]: Thanks I was just being lazy as this css was from my another project.

**Review by xia0m — APPROVED** [review]

**Pull request merged**